client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,576 @@
+				// should be disconnected.
+				// Note that we don't need to update slots, so we drop the node manually here
+				// and not via [`Peer`] interface
+				*node.get_mut() = ConnectionState::NotConnected;
+				let _ = self
+					.to_notifications
+					.unbounded_send(Message::Drop { set_id: self.set_id, peer_id: *node.key() });
+				trace!(
+					target: "peerset",
+					"Disconnecting previously reserved node {} on {:?}.",
+					node.key(), self.set_id
+				);
+			} else {
+				// Otherwise, just count occupied slots for the node.
+				match node.get() {
+					ConnectionState::In => self.num_in += 1,
+					ConnectionState::Out => self.num_out += 1,
+					ConnectionState::NotConnected => {},
+				}
+			}
+		} else {
+			error!(
+				target: "peerset",
+				"Invalid state: reserved node {} not in the list of known nodes.",
+				peer_id,
+			);
+		}
+	}
+
+	fn on_set_reserved_peers(&mut self, peer_ids: HashSet<PeerId>) {
+		// Determine the difference between the current group and the new list.
+		let to_insert = peer_ids.difference(&self.reserved_nodes).cloned().collect::<Vec<_>>();
+		let to_remove = self.reserved_nodes.difference(&peer_ids).cloned().collect::<Vec<_>>();
+
+		for node in to_insert {
+			self.on_add_reserved_peer(node);
+		}
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		self.reserved_only = reserved_only;
+
+		if reserved_only {
+			// Disconnect all non-reserved peers.
+			for peer_id in self.connected_peers().cloned().collect::<Vec<_>>().iter() {
+				self.peer(&peer_id)
+					.into_connected()
+					.expect(
+						"We are enumerating connected peers, therefore the peer is connected; qed",
+					)
+					.disconnect();
+			}
+		} else {
+			// Try connecting to non-reserved peers.
+			self.alloc_slots();
+		}
+	}
+
+	fn on_add_peer(&mut self, peer_id: PeerId) {
+		if self.nodes.insert(peer_id, ConnectionState::NotConnected).is_none() {
+			self.alloc_slots();
+		}
+	}
+
+	fn on_remove_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains(&peer_id) {
+			return
+		}
+
+		self.peer(&peer_id).into_connected().map(|connected_peer| connected_peer.disconnect());
+		self.nodes.remove(&peer_id);
+	}
+
+	fn on_incoming_connection(&mut self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		if self.reserved_only && !self.reserved_nodes.contains(&peer_id) {
+			let _ = self.to_notifications.unbounded_send(Message::Reject(incoming_index));
+			return
+		}
+
+		let not_connected = match self.peer(&peer_id) {
+			// If we're already connected, don't answer, as the docs mention.
+			Peer::Connected(_) => return,
+			Peer::NotConnected(peer) => peer,
+			// Adding incoming peer to our set of peers even before we decide whether to accept
+			// it is questionable, but this is how it was implemented in the original `Peerset`.
```

Comment:
I also think it's questionable and this could be Chesterson's fence type situation but I'm thinking should we remove this `Unknown` abstraction entirely. The end goal for all this refactoring is that the each `ProtocolController` would only have a set of peers it's actively connected to. I know it's probably not achievable in this PR because we still need to keep `Peerset` around for a while longer but peers we're not connected to (and especially unknown peers) should not be stored in `ProtocolController`. Furthermore, `ConnectedPeer` storing a mutable reference to `ProtocolController` makes me think there must be easier way to store peer information and associated states.

I've left another comment below about this issue.

Reply:
I think that sounds good and it should also make the `alloc_slots()` function simpler as we can just collect all entries from both hashmaps and pass that combined hashmap to `Peerset` when querying new candidates